src/cargo/core/compiler/timings.rs

```diff
@@ -0,0 +1,861 @@
+        profile: String,
+    ) -> Timings<'a, 'cfg> {
+        let mut root_map: HashMap<PackageId, Vec<String>> = HashMap::new();
+        for unit in root_units {
+            let target_desc = unit.target.description_named();
+            root_map
+                .entry(unit.pkg.package_id())
+                .or_default()
+                .push(target_desc);
+        }
+        let root_targets = root_map
+            .into_iter()
+            .map(|(pkg_id, targets)| {
+                let pkg_desc = format!("{} {}", pkg_id.name(), pkg_id.version());
+                (pkg_desc, targets)
+            })
+            .collect();
+        let start_str = format!("{}", humantime::format_rfc3339_seconds(SystemTime::now()));
+        let has_report = |what| {
+            config
+                .cli_unstable()
+                .timings
+                .as_ref()
+                .map_or(false, |t| t.iter().any(|opt| opt == what))
+        };
+
+        Timings {
+            config,
+            report_html: has_report("html"),
+            report_info: has_report("info"),
+            report_json: has_report("json"),
+            start: config.creation_time(),
+            start_str,
+            root_targets,
+            profile,
+            total_fresh: 0,
+            total_dirty: 0,
+            unit_times: Vec::new(),
+            active: HashMap::new(),
+            concurrency: Vec::new(),
+        }
+    }
+
+    /// Mark that a unit has started running.
+    pub fn unit_start(&mut self, id: u32, unit: Unit<'a>) {
+        let mut target = if unit.target.is_lib() && unit.mode == CompileMode::Build {
+            // Special case for brevity, since most dependencies hit
+            // this path.
+            "".to_string()
+        } else {
+            format!(" {}", unit.target.description_named())
+        };
+        match unit.mode {
+            CompileMode::Test => target.push_str(" (test)"),
+            CompileMode::Build => {}
+            CompileMode::Check { test: true } => target.push_str(" (check-test)"),
+            CompileMode::Check { test: false } => target.push_str(" (check)"),
+            CompileMode::Bench => target.push_str(" (bench)"),
+            CompileMode::Doc { .. } => target.push_str(" (doc)"),
+            CompileMode::Doctest => target.push_str(" (doc test)"),
+            CompileMode::RunCustomBuild => target.push_str(" (run)"),
+        }
+        let unit_time = UnitTime {
+            unit,
+            target,
+            start: d_as_f64(self.start.elapsed()),
+            duration: 0.0,
+            rmeta_time: None,
+            unlocked_units: Vec::new(),
+            unlocked_rmeta_units: Vec::new(),
+        };
+        assert!(self.active.insert(id, unit_time).is_none());
+    }
+
+    /// Mark that the `.rmeta` file as generated.
+    pub fn unit_rmeta_finished(&mut self, id: u32, unlocked: Vec<&Unit<'a>>) {
+        if let Some(unit_time) = self.active.get_mut(&id) {
```

Comment:
It cannot.  "fresh" units always emit `Message::Finish`, but the timing code only populates its internal "active" map for dirty units.  I've added a comment explaining.

Reply:
Ah ok, failing that could this be restructured to avoid the indentation?

```rust
let unit_time = match self.active.get_mut(&id) {
    Some(unit_time) => time,
    None => return,
};
// ...
```